src/api/c/pinverse.cpp

```diff
@@ -0,0 +1,134 @@
+/*******************************************************
+ * Copyright (c) 2014, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <backend.hpp>
+#include <Array.hpp>
+
+#include <af/array.h>
+#include <af/complex.h>
+#include <af/defines.h>
+#include <af/lapack.h>
+#include <arith.hpp>
+#include <blas.hpp>
+#include <cast.hpp>
+#include <common/ArrayInfo.hpp>
+#include <common/err_common.hpp>
+#include <diagonal.hpp>
+#include <handle.hpp>
+#include <logic.hpp>
+#include <math.hpp>
+#include <select.hpp>
+#include <svd.hpp>
+#include <transpose.hpp>
+
+using af::dim4;
+using namespace detail;
+
+template<typename T>
+Array<T> pinverse_svd(const Array<T> &in)
+{
+    // Moore-Penrose Pseudoinverse
+
+    in.eval();
+    int M = in.dims()[0];
+    int N = in.dims()[1];
+
+    // Compute SVD
+    typedef typename af::dtype_traits<T>::base_type Tr;
+    Array<Tr> sVec = createEmptyArray<Tr>(dim4(min(M, N)));
+    Array<T> u = createEmptyArray<T>(dim4(M, M));
+    Array<T> vT = createEmptyArray<T>(dim4(N, N));
+    svd<T, Tr>(sVec, u, vT, in);
+
+    Array<T> v = transpose(vT, true);
+
+    // Round down small values to zero to avoid large reciprocals later
+    Array<Tr> eps = createValueArray<Tr>(sVec.dims(), scalar<Tr>(1e-6));
+    Array<char> cond = logicOp<Tr, af_lt_t>(sVec, eps, sVec.dims());
+    Array<Tr> sVecRoundOff = createSelectNode<Tr, true>(cond, sVec, 0., sVec.dims());
+
+    // Generate s+
+    Array<Tr> ones = createValueArray<Tr>(sVecRoundOff.dims(), scalar<Tr>(1.));
+    Array<Tr> sVecRecip = arithOp<Tr, af_div_t>(ones, sVecRoundOff, sVecRoundOff.dims());
+    Array<Tr> sPinv = diagCreate<Tr>(sVecRecip, 0);
+
+    // Cast s+ back to original data type for matmul later
+    // (since svd() makes s' type the base type of T)
+    Array<T> sPinvCast = cast<T, Tr>(sPinv);
+
+    Array<T> uT = transpose(u, true);
+
+    // Crop v and u* for final matmul later based on s+'s size, because
+    // sVec produced by svd() has minimal dim length (no extra zeroes).
+    // Thus s+ produced by diagCreate() will have minimal dims as well,
+    // and v could have an extra dim0 or u* could have an extra dim1
+    if (v.dims()[1] > sPinvCast.dims()[0]) {
```

Comment:
By the way, I'm wondering if this is the fastest implementation of what I'm trying to do in this section. At first, actually, I thought of avoiding creating new arrays (`createSubArray()`) since this might be expensive especially when the arrays are big. Instead, a reference to only the needed section of the arrays might be better to use - although this could also have the disadvantage of making the data access pattern jump near the cropped-out section of the arrays during `matmul()`.

Another way could be padding `sVec` with 0's instead to match v and u*'s dims for matmul. This creates a new array once (and a smaller one compared to the other arrays), but will waste multiplication and adding with 0's during matmul.